### PR TITLE
WT-2287: WT_SESSION.rebalance support.

### PR DESCRIPTION
**Pull request by keithbostic**
@agorrod, @michaelcahill: initial WT_SESSION.rebalance support.

One big question (well, aside from whether we want this change at all), is the rebalance method lies between normal processing (which opens an underlying checkpoint/file) and salvage and verify (that don't open an underlying checkpoint/file at all), by opening the checkpoint/file but only loading the root page. It does this because it was easier to use the standard checkpoint/cache-flush mechanisms to create a checkpoint in the underlying file, but I think it's potentially fragile. A better idea or comments would be welcome.

### CODE REVIEW

**Comment by agorrod** [comment]
@keithbostic I generally really like this change. I haven't looked at the details yet.

> One big question, is the rebalance method lies between normal processing and salvage and verify

I'd prefer this to be as close to normal processing as possible. I'd like the semantic to be similar to a checkpoint, except we'd need exclusive access to the live checkpoint, since concurrent updates to this table would be bad. So, like a regular checkpoint except get exclusive access to the handle and ensure there are no open checkpoint handles.

**Comment by keithbostic** [comment]
> So, like a regular checkpoint except get exclusive access to the handle and ensure there are no open checkpoint handles.

I think this is what I've done (but someone should check me carefully, I'm not sure I fully understand the semantics around `WT_DHANDLE_EXCLUSIVE`. (For example, it's unclear to me why we check `WT_BTREE_SPECIAL_FLAGS` in some places and `WT_DHANDLE_EXCLUSIVE` in others?)

I don't think there's a problem with other open checkpoint handles, we only need to make sure we have exclusive access to the live checkpoint.

**Comment by agorrod** [comment]
@keithbostic The change looks generally good to me. Could you add the functionality to ``test/fops`` - that should shake out any handle locking problems.

It would be nice to add some testing into the Python test suite. Something simple that ensures the API is doing something (maybe track a statistic?)

**Comment by keithbostic** [comment]
@agorrod, I added a smoke-test to the Python test suite. I didn't add anything to the fops test, mostly because fops doesn't ever populate its objects and we won't rebalance empty trees, so I'd have to add enough infrastructure to fops so we can insert key/value pairs into trees to make that work.

Let me know if you think it's worth doing that work and I will.

**Comment by agorrod** [comment]
> I didn't add anything to the fops test, mostly because fops doesn't ever populate its objects and we won't rebalance empty trees

Closing the loop after our conversation. I think it's worth adding rebalance to fops without adding content to the tables. I want to get confidence that rebalance operations don't cause locking problems when other schema operations happen in parallel.

**Comment by keithbostic** [comment]
@agorrod, I've pushed the changes to add WT_SESSION.rebalance to the fops test program.

File: src/btree/bt_rebalance.c [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+	WT_PAGE *root;				/* Created root page */
+
+	uint8_t	 type;				/* Internal page type */
+
+#define	WT_REBALANCE_PROGRESS_INTERVAL	100
+	uint64_t progress;			/* Progress counter */
+
+	WT_ITEM *tmp1;				/* Temporary buffers */
+	WT_ITEM *tmp2;
+} WT_RSTUFF;
```
Comment (agorrod) [inline src/btree/bt_rebalance.c]: I know it's local to the file and matches salvage/verify, but I don't like the name. Could we call it ``WT_REBALANCE_STUFF``? It's not used in line-length critical places.

File: src/btree/bt_rebalance.c [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+}
+
+/*
+ * __rebalance_leaf_append --
+ *	Add a new entry to the list of leaf pages.
+ */
+static int
+__rebalance_leaf_append(WT_SESSION_IMPL *session,
+    const uint8_t *key, size_t key_len,
+    uint64_t recno,
```
Comment (agorrod) [inline src/btree/bt_rebalance.c]: Line wrapping looks imperfect here.

File: src/include/wiredtiger.in [hunk 3]
```diff
@@ -1329,6 +1329,20 @@ struct __wt_session {
 	 */
 	int __F(log_printf)(WT_SESSION *session, const char *fmt, ...);
 
+#if !defined(DOXYGEN)
```
Comment (agorrod) [inline src/include/wiredtiger.in]: I'm inclined to make this documented. Having a utility and code in the examples makes me think we are better to document the functionality than attempt to hide it.

File: src/btree/bt_rebalance.c (hunk 1 above)
Comment (keithbostic) [inline src/btree/bt_rebalance.c]: > Could we call it WT_REBALANCE_STUFF?

OK, pushed.

File: src/btree/bt_rebalance.c (hunk 2 above)
Comment (keithbostic) [inline src/btree/bt_rebalance.c]: > Line wrapping looks imperfect here.

Fixed.

File: src/include/wiredtiger.in (hunk 3 above)
Comment (keithbostic) [inline src/include/wiredtiger.in]: > I'm inclined to make this documented.

FBM; whatever you and @michaelcahill decide.

**Comment by michaelcahill** [comment]
@keithbostic, I was asked about whether this should remain undocumented.  Earlier I was trying to have a bet each way.

Since rebalance appears in the utility and the examples, I think we should make it documented in `wiredtiger.in` and mention it in the reference guide (probably along with `compact` in "Advanced topics").

File: src/btree/bt_rebalance.c [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+	WT_REF **refp;
+	uint32_t i, leaf_next;
+
+	btree = S2BT(session);
+
+	/*
+	 * There's a limit to the number of pages we can rebalance: the number
+	 * of elements on a page is a 4B quantity and it's technically possible
+	 * there could be more pages than that in a tree.
+	 */
```
Comment (agorrod) [inline src/btree/bt_rebalance.c]: To ensure I understand - will this rebalance operation create a single level btree? i.e: It only ever creates a single internal page.

Comment (keithbostic) [inline src/btree/bt_rebalance.c]: @agorrod 
> will this rebalance operation create a single level btree? i.e: It only ever creates a single internal page.

Rebalance creates a single in-memory internal page, but reconciliation splits that page until it's under the maximum internal page size on disk, reconciliation creates multiple Btree levels as needed. (It's the `rec_root_write` code inside reconciliation.)

The problem with that, of course, is rebalance (along with bulk-load and salvage), requires the entire tree's internal key/value information fit in memory -- we're probably going to need to fix that at some point.

**Comment by keithbostic** [comment]
@michaelcahill 
> Since rebalance appears in the utility and the examples, I think we should make it documented in wiredtiger.in and mention it in the reference guide (probably along with compact in "Advanced topics").

Done and pushed.

**Comment by agorrod** [comment]
Thanks @keithbostic. lgtm I'll merge.

File: src/btree/bt_rebalance.c (hunk 4 above)
Comment (erbenmo) [inline src/btree/bt_rebalance.c]: Looks like the reconciliation is invoked by the checkpoint that we do when we close the connection. I am wondering if that checkpoint always going to be executed. I remember that checkpoint has a "sync" config which user can control to whether they want to do a checkpoint or not when the connection is closed...

Comment (michaelcahill) [inline src/btree/bt_rebalance.c]: @erbenmo, since `rebalance` is an exclusive operation, the tree will be flushed to disk when the operation completes.  That is independent of checkpoints or `WT_CONNECTION::close`.
